Index: exec/objdb.c
===================================================================
--- exec/objdb.c	(revision 2944)
+++ exec/objdb.c	(revision 2945)
@@ -1539,8 +1539,10 @@
 	char *str;
 	ret = object_key_iter_typed (parent_object_handle,
 		(char**)key_name, value, value_len, &t);
-	str = *key_name;
-	*key_len = strlen(str);
+	if (!ret) {
+		str = *key_name;
+		*key_len = strlen(str);
+	}
 	return ret;
 }
 
