Index: exec/sync.c
===================================================================
--- exec/sync.c	(revision 2999)
+++ exec/sync.c	(revision 3000)
@@ -479,8 +479,8 @@
 	memcpy (my_member_list, member_list, member_list_entries * sizeof (unsigned int));
 	my_member_list_entries = member_list_entries;
 
+	sync_aborted ();
 	if (sync_processing && sync_callbacks.sync_abort != NULL) {
-		sync_aborted ();
 		sync_callbacks.sync_abort ();
 		sync_callbacks.sync_activate = NULL;
 	}
