Index: exec/totemsrp.c
===================================================================
--- exec/totemsrp.c	(revision 3012)
+++ exec/totemsrp.c	(revision 3013)
@@ -3740,10 +3740,6 @@
 	}
 #endif
 
-        if (srp_addr_equal (&mcast_header.system_from, &instance->my_id) == 0) {
-		cancel_token_retransmit_timeout (instance);
-	}
-
 	/*
 	 * If the message is foreign execute the switch below
 	 */
