Index: corosync.spec.in
===================================================================
--- corosync.spec.in	(revision 2966)
+++ corosync.spec.in	(revision 2967)
@@ -107,6 +107,8 @@
 %{_libexecdir}/lcrso/vsf_ykd.lcrso
 %dir %{_localstatedir}/lib/corosync
 %{_mandir}/man8/corosync_overview.8*
+%{_mandir}/man8/corosync.8*
+%{_mandir}/man8/corosync-blackbox.8*
 %{_mandir}/man8/corosync-objctl.8*
 %{_mandir}/man8/corosync-keygen.8*
 %{_mandir}/man8/corosync-cfgtool.8*
Index: man/Makefile.am
===================================================================
--- man/Makefile.am	(revision 2966)
+++ man/Makefile.am	(revision 2967)
@@ -39,7 +39,9 @@
 dist_man_MANS = \
 	corosync.conf.5 \
 	confdb_overview.8 \
+	corosync.8 \
 	corosync-objctl.8 \
+	corosync-blackbox.8 \
 	corosync-keygen.8 \
 	corosync-cfgtool.8 \
 	corosync-cpgtool.8 \
